Convert project to Python module

carpemonf opened the pull request.

carpemonf committed:
Converting DeepCFD into a module

carpemonf committed:
FIxing typo in README

carpemonf committed:
Fixing torch package versions

carpemonf committed:
Removing matplotlib unneeded package

carpemonf committed:
Fixing help options message

carpemonf committed:
Fixing unit of pressure in visualization script

mdribeiro commented on src/deepcfd/__main__.py:
```diff
@@ -0,0 +1,251 @@
+                "\n    -l <learning-rate>  learning rate (default: 0.001)"
+                "\n    -e <epochs>  number of epochs (default: 1000)"
+                "\n    -b <batch-size>  training batch size (default: 32)"
+                "\n    -p <patience>  number of epochs for early stopping (default: 300)\n"
```
As mentioned in previous comment, if you accept the suggestion to include the device and visualize options, please remember do at the info here.

mdribeiro commented on src/deepcfd/__main__.py:
```diff
@@ -0,0 +1,251 @@
+    state_dict["kernel_size"] = options["kernel_size"]
+    state_dict["architecture"] = options["net"]
+    
+    torch.save(state_dict, options["output"])
```
If you accept the visualize option, I would suggest adding it here after saving the model.

mdribeiro reviewed (commented):
Hi Carlos,
Thank you very much for sending this pull request. Awesome job!
I just had a few suggestions and questions. Feel free to take a look.
Once you're ready, please let me know and provide I commit message. I will be happy to merge it to the master branch. At some point, it may make sense to setup a CI/CD pipeline to make sure things are going forward correctly, but it may take me some more time to do.

carpemonf commented on scripts/visualize_data.py:
```diff
@@ -0,0 +1,145 @@
+import pickle
+import matplotlib
+import numpy as np
+from cfdonnx.models.torch.UNetEx import UNetEx
```
Yes, sorry it slipped through the final version.

carpemonf commented on scripts/visualize_data.py:
```diff
@@ -0,0 +1,145 @@
+
+matplotlib.use('tkagg')
+
+model_file="checkpoint.pt"
```
Yes, should be model_filename.

carpemonf commented on README.md:
```diff
@@ -6,6 +6,48 @@ Paper: https://arxiv.org/abs/2004.08826
+
+Example:
+
+python3 -m deepcfd \
```
Hi Mateus,

Thank you for your comments. I'm finishing the new commits with your suggestions, but wanted to check with you first about the visualization before applying the changes.

I decoupled the visualization functions from the deepcfd module as I had the idea that the module should be focused on performing the training task in a generic way. Also, it removes the matplotlib as a dependency of the module as this can be done separately. For these reasons,  I moved the visualization to a separate script to serve as a template that anyone can adapt it to build their specific plots.

The `visualization_data.py` monitors the checkpoint and log file and updates the plot every time a new checkpoint is written. In that way, one can see how the training evolves and be also used to visualize training processes happening in the background in a remote server like an HPC. It allows a better customization of the plots as moving the visualize function into the module makes this plot a default. With this approach people could install deepcfd as a module and take the visualization script as a template in their projects/folders. You mentioned that there were some issues using it with the toy dataset. Could you please share the error? I think I tried that dataset too with this script, but maybe I made some changes later and it needs some adjustments.

I think that the `visualize` flag you suggested is great and makes sense, as it gives users the option to run the program interactively or not. I have no strong feelings about the above, but I wanted to let you know my reasoning behind it. Please let me know your thoughts and I will submit the final version.

Cheers!

mdribeiro replied:
Hi @carpemonf , thanks for clarifying it! Regarding the problems I had, I was able to solve it. I was passing the wrong checkpoint to the script. The only issue that I had is that the plots were not rendering very well in the plot, so I transposed the tensors with `.T` and now everything works fine.

Now that I understand it, I agree with you and I think it makes sense what you had in mind before. I like how easy it is to generate the plots based on the checkpoints, and as you mentioned, it is nice to decouple the matplotlib dependency from the rest of the module.

I leave it up to you to add the visualize flag and have the option to run the program interactively. I personally think it is nice, since it gives the user a chance to actually see what is happening on the run. But I have no strong feelings about it either. ;)

Thanks a lot! Looking forward to your new commits. Once you're ready, let me know so I can take a look and merge it.

carpemonf replied:
Hi @mdribeiro,

Please check the last couple of commits. I think they include everything you mentioned, but let me know if not!

There's now a device option that can be set to `cpu`, `cuda`, `cuda:0`, `cuda:0,cuda:1` and so on. If there's no such option, and `cuda` is available, `cuda` will be the choice as it was before.

I see the value for users to have everything in one go, including the data visualization, so I added the visualize flag. It's disabled by default but can be activated with the `--visualize` flag, e.g. `python3 -m deepcfd --visualize`. If `--visualize` or `-v` are not specified, the plots are not displayed.

Finally, I have updated the README for pointing the module's installation to your `mdribeiro:master`. Also, notice that it's all set-up for publication to PyPI if you decide to do that. In that case, one could install it as `pip3 install deepcfd`.

Cheers!

mdribeiro replied:
Hi @carpemonf,

Thank you so much for putting in the great work! I think your changes look great and make the project more suitable to be deployed in other scenarios and it will make it much easier for me to work on future extensions.

I just added a few minor corrections in the README (you swapped the help messages between the --device and --net options), and I also took the liberty to add your name as author in the setup.py file, since you did a massive restructuring.

I will do the merge now. As I told you before, I will be working on professionalizing this project by addings tests and documentation. Next step will be to add a few tests. It would be great if you could review the new changes.

Best,
Mateus

carpemonf replied:
Hi @mdribeiro,

Not at all, glad to be of help for your project. Thanks for adding me as author, indeed!

Your work on adding tests and documentation is really good news. I will be glad to review them. Looking forward to it.

carpemonf commented on scripts/visualize_data.py:
```diff
@@ -0,0 +1,145 @@
+matplotlib.use('tkagg')
+
+model_file="checkpoint.pt"
+data_x_file="DeepCFD/flowAroundObstaclesX.pkl"
```
Sure, I was using another dataset and forgot to bring back the names to your original dataset. However, I think it's better to name them `dataX.pkl` and `dataY.pkl` and avoid the relative paths. For instance, it could be called from the root as `python3 scripts/visualize_data.py`.

carpemonf committed:
Adding device option to set cpu or gpu

carpemonf committed:
Bringing back the visualize function

A  flag is added in the main program for using it

carpemonf committed:
Fixing variable names in the visualization template script

carpemonf committed:
Updating README with updated module options

The installation instructions have been fixed as well for pointing to the original repo
